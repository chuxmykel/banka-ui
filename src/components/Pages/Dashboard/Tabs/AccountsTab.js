import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import moment from 'moment';
import Loader from 'react-loader-spinner';
import Tab from '@Dashboard/Common/Tab/Tab';
import { getAccounts, getAccountTransactions } from '@Actions/accountActions';
import { openModal } from '@Actions/uiActions';
import CreateAccount from '@Components/Forms/CreateAccount/CreateAccount';

class AccountTab extends Component {
  constructor(props) {
    super(props);
    this.state = {
      active: props.active,
      showTransactionPane: false,
    };
  }

  componentDidUpdate = (prevProps) => {
    const { active, getUserAccounts } = this.props;
    if (prevProps.active !== active) {
      if (active === true) getUserAccounts();
      this.setState({
        active,
        showTransactionPane: false,
      });
    }
  }

  render = () => {
    const { active, showTransactionPane } = this.state;
    const {
      accounts,
      getTransactions,
      transactions,
      open,
      loading,
    } = this.props;

    const loader = <Loader type="ThreeDots" color="#888888" height={50} width={100} />;
    const accountData = accounts.map(account => (
      <tr key={account.id}>
        <td>{account.type}</td>
        <td className={`${account.status}-acct`}>{account.status}</td>
        <td>{account.accountNumber}</td>
        <td>{`NGN ${account.balance}`}</td>
        <td>
          <button
            type="button"
            className="acct-btn"
            onClick={() => {
              this.setState({
                showTransactionPane: true,
              });
              getTransactions(account.accountNumber);
            }}
          >
            Transactions
          </button>
        </td>
      </tr>
    ));

    const mapTransactions = (trans) => {
      let data = [];
      if (trans === undefined) {
        data = [(
          <tr>
            <td colSpan="4">
              Selected account has no transactions.
              Visit one of our branches to carry out transactions.
            </td>
          </tr>
        )];
      } else {
        data = trans.map((transaction) => {
          const { type } = transaction;
          return (
            <tr key={transaction.id}>
              <td>
                <i
                  className={`fas fa-long-arrow-alt-${type === 'credit' ? 'up' : 'down'} ${type}`}
                />
              </td>
              <td>{`NGN ${transaction.amount}`}</td>
              <td>{`${moment(transaction.createdOn).format('L')}`}</td>
              <td>{`${moment(transaction.createdOn).format('LT')}`}</td>
            </tr>
          );
        });
      }
      return (
        <tbody>
          {data}
        </tbody>
      );
    };
    const transactionData = mapTransactions(transactions);
    return (
      <Tab active={active}>
        <div className="account-tab">
          <div className="name-card">
            accounts
          </div>
          <div className="account-details">
            <table>
              <thead>
                <tr>
                  <th>Account Type</th>
                  <th>Account Status</th>
                  <th>Account Number</th>
                  <th>Account Balance</th>
                  <th />
                </tr>
              </thead>
              <tbody>
                {accountData}
              </tbody>
            </table>
          </div>
          {loading ? loader : (
            <div
              className={
                `account-details transaction${showTransactionPane ? '' : ' hidden'}`
              }
            >
              <h2 className="transaction-header">Transactions</h2>
              <table>
                <thead>
                  <tr>
                    <th>
                      <i className="fas fa-long-arrow-alt-up credit" />
                      <i className="fas fa-long-arrow-alt-down debit" />
                    </th>
                    <th>Amount</th>
                    <th>Date</th>
                    <th>Time</th>
                  </tr>
                </thead>
                {transactionData}
              </table>
            </div>
          )}
        </div>
        <div>
          <button onClick={() => open('createAccount')} className="create" type="button">+</button>
        </div>
        <CreateAccount />
      </Tab>
    );
  }
}

AccountTab.propTypes = {
  active: PropTypes.bool.isRequired,
  getUserAccounts: PropTypes.func.isRequired,
  accounts: PropTypes.array.isRequired,
  transactions: PropTypes.array,
  getTransactions: PropTypes.func.isRequired,
  open: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
};

AccountTab.defaultProps = {
  transactions: undefined,
};

const mapStateToProps = state => ({
  accounts: state.accounts.accounts,
  transactions: state.accounts.transactions,
  loading: state.ui.loading,
});

const mapDispatchToProps = dispatch => ({
  open: modal => dispatch(openModal(modal)),
  getUserAccounts: () => dispatch(getAccounts()),
  getTransactions: accNum => dispatch(getAccountTransactions(accNum)),
});

export default connect(mapStateToProps, mapDispatchToProps)(AccountTab);
