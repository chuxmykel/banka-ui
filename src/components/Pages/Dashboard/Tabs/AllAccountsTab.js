/* eslint-disable jsx-a11y/label-has-associated-control */
/* eslint-disable jsx-a11y/label-has-for */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Loader from 'react-loader-spinner';
import Tab from '@Dashboard/Common/Tab/Tab';
import DialogueBox from '@Common/DialogueBox/DialogueBox';
import NewTransaction from '@Components/Forms/NewTransaction/NewTransaction';
import {
  getAllUsersAccounts,
  changeAccountStatus,
  deleteAccount,
} from '@Actions/accountActions';
import { openModal } from '@Actions/uiActions';

class AllAccountsTab extends Component {
  constructor(props) {
    super(props);
    this.state = {
      active: props.active,
      activeDialogue: '',
      dialogueMessage: '',
      dialogueData: {},
    };
  }

  componentDidUpdate = (prevProps) => {
    const { active, getUserAccounts } = this.props;
    if (prevProps.active !== active) {
      if (active === true) getUserAccounts();
      this.setState({
        active,
      });
    }
  }

  handleChange = (account) => {
    const { open } = this.props;
    open('DialogueBox');
    this.setState({
      activeDialogue: 'edit',
      dialogueMessage: `Are you sure you want to
      ${account.status === 'active' ? 'deactivate' : 'activate'}
      this account? It can be undone later by toggling the switch`,
      dialogueData: account,
    });
  }

  handleClick = (accNumber) => {
    const { open } = this.props;
    open('DialogueBox');
    this.setState({
      activeDialogue: 'delete',
      dialogueMessage: 'Are you sure you want to delete this account?',
      dialogueData: accNumber,
    });
  }

  render = () => {
    const {
      active,
      activeDialogue,
      dialogueMessage,
      dialogueData,
    } = this.state;
    const {
      accounts,
      editAccount,
      deleteAcct,
      loading,
      open,
      cashier,
    } = this.props;

    const loader = <Loader type="ThreeDots" color="#888888" height={50} width={100} />;
    const accountData = accounts.map(account => (
      <tr key={account.id}>
        <td>{`${account.firstName} ${account.lastName.charAt(0)}.`}</td>
        <td>{account.accountNumber}</td>
        <td>{account.type}</td>
        <td className={`${account.status}-acct`}>{account.status}</td>
        <td>
          <label className="switch">
            <input checked={account.status === 'active'} type="checkbox" onChange={() => this.handleChange(account)} />
            <span className="slider round" />
          </label>
        </td>
        <td>{account.balance}</td>
        <td>
          <button
            type="button"
            className="delete-btn"
            onClick={() => this.handleClick(account.accountNumber)}
          >
            delete
          </button>
        </td>
      </tr>
    ));

    return (
      <Tab active={active}>
        <div className="account-tab">
          <div className="name-card">
            all user&apos;s accounts
          </div>
          <div className="account-details admin-acct-details">
            <table>
              <thead>
                <tr>
                  <th>Owner</th>
                  <th>Acct. Number</th>
                  <th>Type</th>
                  <th>Status</th>
                  <th />
                  <th>Balance (â‚¦)</th>
                  <th />
                </tr>
              </thead>
              <tbody>
                {accountData}
              </tbody>
            </table>
          </div>
        </div>
        <DialogueBox
          agreeText={dialogueData.status === 'active' ? 'deactivate' : 'activate'}
          activeDialogue={activeDialogue === 'edit'}
          agree={() => editAccount(dialogueData)}
        >
          {loading ? loader : (<p>{dialogueMessage}</p>)}
        </DialogueBox>
        <DialogueBox
          agreeText="delete"
          activeDialogue={activeDialogue === 'delete'}
          agree={() => deleteAcct(dialogueData)}
        >
          {loading ? loader : (<p>{dialogueMessage}</p>)}
        </DialogueBox>
        {cashier && (
          <div>
            <button
              onClick={() => open('newTransaction')}
              className="create"
              type="button"
            >
              +
            </button>
          </div>
        )}
        <NewTransaction />
      </Tab>
    );
  }
}

AllAccountsTab.propTypes = {
  active: PropTypes.bool.isRequired,
  getUserAccounts: PropTypes.func.isRequired,
  editAccount: PropTypes.func.isRequired,
  deleteAcct: PropTypes.func.isRequired,
  accounts: PropTypes.array.isRequired,
  open: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
  cashier: PropTypes.bool.isRequired,
};

const mapStateToProps = state => ({
  accounts: state.accounts.accounts,
  loading: state.ui.loading,
  cashier: state.auth.user.type === 'staff' && state.auth.user.isAdmin === false,
});

const mapDispatchToProps = dispatch => ({
  open: modal => dispatch(openModal(modal)),
  getUserAccounts: () => dispatch(getAllUsersAccounts()),
  editAccount: account => dispatch(changeAccountStatus(account)),
  deleteAcct: accountNumber => dispatch(deleteAccount(accountNumber)),
});

export default connect(mapStateToProps, mapDispatchToProps)(AllAccountsTab);
