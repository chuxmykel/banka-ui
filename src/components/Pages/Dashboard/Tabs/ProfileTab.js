import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Tab from '@Dashboard/Common/Tab/Tab';

class ProfileTab extends Component {
  constructor(props) {
    super(props);
    this.state = {
      active: props.active,
    };
  }

  componentDidUpdate(prevProps) {
    const { active } = this.props;
    if (prevProps.active !== active) {
      this.setState({
        active,
      });
    }
  }

  render = () => {
    const {
      user: {
        firstName,
        lastName,
        email,
        type,
        isAdmin,
      },
    } = this.props;
    const { active } = this.state;
    const role = isAdmin ? 'Admin' : 'Cashier';
    return (
      <Tab active={active}>
        <div className="profile-tab">
          <div className="name-card">
            {`${firstName} ${lastName}`}
          </div>
          <div className="profile-details">
            <table>
              <tbody>
                <tr>
                  <th>{type !== 'client' ? 'ROLE' : 'BVN'}</th>
                  <td>{type !== 'client' ? role : '*******************'}</td>
                </tr>
                <tr>
                  <th>EMAIL</th>
                  <td>{email}</td>
                </tr>
                <tr>
                  <th>PHONE</th>
                  <td>+234(0)7060854773</td>
                </tr>
                <tr>
                  <th>ADDRESS</th>
                  <td>No. 32, Australopitecus Avenue, CIPE tower.</td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </Tab>
    );
  };
}

ProfileTab.propTypes = {
  user: PropTypes.object.isRequired,
  active: PropTypes.bool.isRequired,
};

const mapStateToProps = state => ({
  user: state.auth.user,
});

export default connect(mapStateToProps, null)(ProfileTab);
