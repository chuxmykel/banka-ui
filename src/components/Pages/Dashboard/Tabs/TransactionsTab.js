import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Tab from '@Dashboard/Common/Tab/Tab';

class TransactionTab extends Component {
  constructor(props) {
    super(props);
    this.state = {
      active: props.active,
    };
  }

  componentDidUpdate = (prevProps) => {
    const { active } = this.props;
    if (prevProps.active !== active) {
      this.setState({
        active,
      });
    }
  }

  render = () => {
    const { active } = this.state;
    return (
      <Tab active={active}>
        <div className="name-card">
          transactions
        </div>
      </Tab>
    );
  }
}

TransactionTab.propTypes = {
  active: PropTypes.bool.isRequired,
};

const mapStateToProps = state => ({
  user: state.auth.user,
});

export default connect(mapStateToProps, null)(TransactionTab);
