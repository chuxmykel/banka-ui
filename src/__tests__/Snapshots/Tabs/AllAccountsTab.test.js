import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';
import { AllAccountsTab } from '@Pages/Dashboard/Tabs/AllAccountsTab';

it('Renders correctly', () => {
  const wrapper = shallow(
    <AllAccountsTab
      active
      getUserAccounts={jest.fn()}
      editAccount={jest.fn()}
      deleteAcct={jest.fn()}
      accounts={[]}
      open={jest.fn()}
      loading={false}
      cashier={false}
    />,
  );
  expect(toJson(wrapper)).toMatchSnapshot();
});
