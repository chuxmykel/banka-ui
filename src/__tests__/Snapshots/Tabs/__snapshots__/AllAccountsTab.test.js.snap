// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
<Tab
  active={true}
>
  <div
    className="account-tab"
  >
    <div
      className="name-card"
    >
      all user's accounts
    </div>
    <div
      className="account-details admin-acct-details"
    >
      <table>
        <thead>
          <tr>
            <th>
              Owner
            </th>
            <th>
              Acct. Number
            </th>
            <th>
              Type
            </th>
            <th>
              Status
            </th>
            <th />
            <th>
              Balance (â‚¦)
            </th>
            <th />
          </tr>
        </thead>
        <tbody />
      </table>
    </div>
  </div>
  <Connect(DialogueBox)
    activeDialogue={false}
    agree={[Function]}
    agreeText="activate"
  >
    <p />
  </Connect(DialogueBox)>
  <Connect(DialogueBox)
    activeDialogue={false}
    agree={[Function]}
    agreeText="delete"
  >
    <p />
  </Connect(DialogueBox)>
  <Connect(NewTransaction) />
</Tab>
`;
